File: jars/src/org/niatahl/tahlan/hullmods/KnightRefit.java
package org.niatahl.tahlan.hullmods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.combat.*;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;

import java.awt.*;

import static org.niatahl.tahlan.utils.Utils.txt;
import static org.niatahl.tahlan.utils.Afterimage.renderCustomAfterimage;

public class KnightRefit extends BaseHullMod {


    public static final float ARMOR_MALUS_FRIGATE = 50f;
    public static final float ARMOR_MALUS_DESTROYER = 100f;
    public static final float ARMOR_MALUS_CRUISER = 200f;
    public static final float ARMOR_MALUS_CAPITAL = 300f;

    public static final float SUPPLIES_MULT = 50f;

    public static final float OVERDRIVE_TRIGGER_PERCENTAGE = 0.3f;
    public static final float OVERDRIVE_TIME_MULT = 30f;

    public static final float TIME_MULT = 10f;
    private static final Color AFTERIMAGE_COLOR = new Color(133, 126, 116, 90);
    private static final float AFTERIMAGE_THRESHOLD = 0.4f;

    private static final Color OVERDRIVE_ENGINE_COLOR = new Color(255, 44, 0);
    private static final Color OVERDRIVE_GLOW_COLOR = new Color(255, 120, 16);
    private static final Color OVERDRIVE_JITTER_COLOR = new Color(255, 63, 0, 20);
    private static final Color OVERDRIVE_JITTER_UNDER_COLOR = new Color(255, 63, 0, 80);

    private static final float SO_MALFUNCTION_PROB = 0.03f;

    private float fadeOut = 1f;
    private static final String ke_id = "tahlan_KnightRefitID";

    private final String INNERLARGE = "graphics/tahlan/fx/tahlan_tempshield.png";
    private final String OUTERLARGE = "graphics/tahlan/fx/tahlan_tempshield_ring.png";

    @Override
    public void applyEffectsBeforeShipCreation(HullSize hullSize, MutableShipStatsAPI stats, String id) {

        switch (hullSize) {
            case FRIGATE:
                stats.getArmorBonus().modifyFlat(ke_id, -ARMOR_MALUS_FRIGATE);
                break;
            case DESTROYER:
                stats.getArmorBonus().modifyFlat(ke_id, -ARMOR_MALUS_DESTROYER);
                break;
            case CRUISER:
                stats.getArmorBonus().modifyFlat(ke_id, -ARMOR_MALUS_CRUISER);
                break;
            case CAPITAL_SHIP:
                stats.getArmorBonus().modifyFlat(ke_id, -ARMOR_MALUS_CAPITAL);
        }

        stats.getSuppliesPerMonth().modifyPercent(ke_id, SUPPLIES_MULT);
        stats.getCRLossPerSecondPercent().modifyMult(ke_id, 2f);

    }

    @Override
    public void applyEffectsAfterShipCreation(ShipAPI ship, String id) {
        if (ship.getVariant().hasHullMod("safetyoverrides")) {
            ship.getMutableStats().getWeaponMalfunctionChance().modifyFlat(id, SO_MALFUNCTION_PROB);
            ship.getMutableStats().getEngineMalfunctionChance().modifyFlat(id, SO_MALFUNCTION_PROB);
            ship.getMutableStats().getCriticalMalfunctionChance().modifyFlat(id, SO_MALFUNCTION_PROB/2);
        }
        ship.getShield().setRadius(ship.getShieldRadiusEvenIfNoShield(), INNERLARGE, OUTERLARGE);
    }

    @Override
    public void advanceInCombat(ShipAPI ship, float amount) {

        //don't run while paused because duh
        if (Global.getCombatEngine().isPaused()) {
            return;
        }

        //The Great Houses are actually timelords
        boolean player = ship == Global.getCombatEngine().getPlayerShip();

        if ( !ship.isAlive() || ship.isPiece() ) {
            return;
        }

        ship.setJitterShields(false);

        if (ship.getSystem() != null) {
            if (!ship.getSystem().isActive() && !ship.getFluxTracker().isOverloadedOrVenting()) {
                if (ship.getHitpoints() <= ship.getMaxHitpoints() * OVERDRIVE_TRIGGER_PERCENTAGE || ship.getVariant().getHullMods().contains("tahlan_forcedoverdrive")) {

                    if (player) {
                        ship.getMutableStats().getTimeMult().modifyPercent(ke_id, OVERDRIVE_TIME_MULT);
                        Global.getCombatEngine().getTimeMult().modifyPercent(ke_id, 1f / OVERDRIVE_TIME_MULT);
                    } else {
                        ship.getMutableStats().getTimeMult().modifyPercent(ke_id, OVERDRIVE_TIME_MULT);
                        Global.getCombatEngine().getTimeMult().unmodify(ke_id);
                    }

                    if (!(ship.getOriginalOwner() == -1)) {

                        ship.getEngineController().fadeToOtherColor(this, OVERDRIVE_ENGINE_COLOR, null, 1f, 0.7f);
                        ship.setJitter(ke_id, OVERDRIVE_JITTER_COLOR, 0.5f, 3, 5f);
                        ship.setJitterUnder(ke_id, OVERDRIVE_JITTER_UNDER_COLOR, 0.5f, 20, 10f);

                    }

                    if (player) {
                        Global.getCombatEngine().maintainStatusForPlayerShip(ke_id, "graphics/icons/hullsys/temporal_shell.png", "Temporal Overdrive", "Timeflow at 130%", false);
                    }

                    fadeOut  = 2f;

                } else {

                    if (player) {
                        ship.getMutableStats().getTimeMult().modifyPercent(ke_id, TIME_MULT);
                        Global.getCombatEngine().getTimeMult().modifyPercent(ke_id, 1f / TIME_MULT);
                        Global.getCombatEngine().maintainStatusForPlayerShip(ke_id, "graphics/icons/hullsys/temporal_shell.png", "Temporal Field", "Timeflow at 110%", false);
                    } else {
                        ship.getMutableStats().getTimeMult().modifyPercent(ke_id, TIME_MULT);
                        Global.getCombatEngine().getTimeMult().unmodify(ke_id);
                    }

                    fadeOut = 1f;

                }

                ship.getMutableStats().getDynamic().getStat("tahlan_KRAfterimageTracker").modifyFlat("tahlan_KRAfterimageTrackerNullerID", -1);
                ship.getMutableStats().getDynamic().getStat("tahlan_KRAfterimageTracker").modifyFlat("tahlan_KRAfterimageTrackerID",
                        ship.getMutableStats().getDynamic().getStat("tahlan_KRAfterimageTracker").getModifiedValue() + amount);
                if (ship.getMutableStats().getDynamic().getStat("tahlan_KRAfterimageTracker").getModifiedValue() > AFTERIMAGE_THRESHOLD) {

                    renderCustomAfterimage(ship,AFTERIMAGE_COLOR,fadeOut);

                    ship.getMutableStats().getDynamic().getStat("tahlan_KRAfterimageTracker").modifyFlat("tahlan_KRAfterimageTrackerID",
                            ship.getMutableStats().getDynamic().getStat("tahlan_KRAfterimageTracker").getModifiedValue() - AFTERIMAGE_THRESHOLD);

                }

            } else {
                ship.getMutableStats().getTimeMult().unmodify(ke_id);
                Global.getCombatEngine().getTimeMult().unmodify(ke_id);
            }
        }




    }


    //Built-in only
    @Override
    public boolean isApplicableToShip(ShipAPI ship) {
        return false;
    }

    public String getDescriptionParam(int index, HullSize hullSize) {
        if (index == 0) return txt("hmd_KassEng1");
        if (index == 1) return "" + (int)TIME_MULT + txt("%");
        if (index == 2)
            return "" + (int) ARMOR_MALUS_FRIGATE + txt("/") + (int) ARMOR_MALUS_DESTROYER + txt("/") + (int) ARMOR_MALUS_CRUISER + txt("/") + (int) ARMOR_MALUS_CAPITAL;
        if (index == 3) return "" + (int)SUPPLIES_MULT + txt("%");
        if (index == 4) return "" + Math.round(OVERDRIVE_TRIGGER_PERCENTAGE * 100f) + txt("%");
        if (index == 5) return "" + (int)OVERDRIVE_TIME_MULT + txt("%");
        if (index == 6) return txt("hmd_KassEng2");
        if (index == 7) return txt("hmd_KassEng3");
        return null;
    }
}

File: jars/src/org/niatahl/tahlan/hullmods/KnightRefit.kt
package org.niatahl.tahlan.hullmods

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.combat.BaseHullMod
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.ShipAPI.HullSize
import org.niatahl.tahlan.plugins.CustomRender
import org.niatahl.tahlan.utils.Afterimage.renderCustomAfterimage
import org.niatahl.tahlan.utils.Utils
import java.awt.Color
import kotlin.math.roundToInt

class KnightRefit : BaseHullMod() {
    private var fadeOut = 1f
    private val INNERLARGE = "graphics/tahlan/fx/tahlan_tempshield.png"
    private val OUTERLARGE = "graphics/tahlan/fx/tahlan_tempshield_ring.png"
    override fun applyEffectsBeforeShipCreation(hullSize: HullSize, stats: MutableShipStatsAPI, id: String) {
        when (hullSize) {
            HullSize.FRIGATE -> stats.armorBonus.modifyFlat(ke_id, -ARMOR_MALUS_FRIGATE)
            HullSize.DESTROYER -> stats.armorBonus.modifyFlat(ke_id, -ARMOR_MALUS_DESTROYER)
            HullSize.CRUISER -> stats.armorBonus.modifyFlat(ke_id, -ARMOR_MALUS_CRUISER)
            HullSize.CAPITAL_SHIP -> stats.armorBonus.modifyFlat(ke_id, -ARMOR_MALUS_CAPITAL)
            else -> stats.armorBonus.modifyFlat(ke_id, -ARMOR_MALUS_FRIGATE)
        }
        stats.suppliesPerMonth.modifyPercent(ke_id, SUPPLIES_MULT)
        stats.crLossPerSecondPercent.modifyMult(ke_id, 2f)
    }

    override fun applyEffectsAfterShipCreation(ship: ShipAPI, id: String) {
        if (ship.variant.hasHullMod("safetyoverrides")) {
            ship.mutableStats.weaponMalfunctionChance.modifyFlat(id, SO_MALFUNCTION_PROB)
            ship.mutableStats.engineMalfunctionChance.modifyFlat(id, SO_MALFUNCTION_PROB)
            ship.mutableStats.criticalMalfunctionChance.modifyFlat(id, SO_MALFUNCTION_PROB / 2)
        }
        ship.shield.setRadius(ship.shieldRadiusEvenIfNoShield, INNERLARGE, OUTERLARGE)
    }

    override fun advanceInCombat(ship: ShipAPI, amount: Float) {

        //don't run while paused because duh
        if (Global.getCombatEngine().isPaused) {
            return
        }

        //The Great Houses are actually timelords
        val player = ship === Global.getCombatEngine().playerShip
        if (!ship.isAlive || ship.isPiece) {
            return
        }
        if (ship.hullSpec.baseHullId !in GH_HULLS) {
            ship.mutableStats.peakCRDuration.modifyMult("lol_lmao_even", 0.1f)
            return
        }
        ship.isJitterShields = false
        if (ship.system != null) {
            if (!ship.system.isActive && !ship.fluxTracker.isOverloadedOrVenting) {
                fadeOut = if (ship.hitpoints <= ship.maxHitpoints * OVERDRIVE_TRIGGER_PERCENTAGE || ship.variant.hullMods.contains("tahlan_forcedoverdrive")) {
                    if (player) {
                        ship.mutableStats.timeMult.modifyPercent(ke_id, OVERDRIVE_TIME_MULT)
                        Global.getCombatEngine().timeMult.modifyPercent(ke_id, 1f / OVERDRIVE_TIME_MULT)
                    } else {
                        ship.mutableStats.timeMult.modifyPercent(ke_id, OVERDRIVE_TIME_MULT)
                        Global.getCombatEngine().timeMult.unmodify(ke_id)
                    }
                    if (ship.originalOwner != -1) {
                        ship.engineController.fadeToOtherColor(this, OVERDRIVE_ENGINE_COLOR, null, 1f, 0.7f)
                        ship.setJitter(ke_id, OVERDRIVE_JITTER_COLOR, 0.5f, 3, 5f)
                        ship.setJitterUnder(ke_id, OVERDRIVE_JITTER_UNDER_COLOR, 0.5f, 20, 10f)
                    }
                    if (player) {
                        Global.getCombatEngine()
                            .maintainStatusForPlayerShip(ke_id, "graphics/icons/hullsys/temporal_shell.png", "Temporal Overdrive", "Timeflow at 130%", false)
                    }
                    2f
                } else {
                    if (player) {
                        ship.mutableStats.timeMult.modifyPercent(ke_id, TIME_MULT)
                        Global.getCombatEngine().timeMult.modifyPercent(ke_id, 1f / TIME_MULT)
                        Global.getCombatEngine()
                            .maintainStatusForPlayerShip(ke_id, "graphics/icons/hullsys/temporal_shell.png", "Temporal Field", "Timeflow at 110%", false)
                    } else {
                        ship.mutableStats.timeMult.modifyPercent(ke_id, TIME_MULT)
                        Global.getCombatEngine().timeMult.unmodify(ke_id)
                    }
                    1f
                }
                ship.mutableStats.dynamic.getStat("tahlan_KRAfterimageTracker").modifyFlat("tahlan_KRAfterimageTrackerNullerID", -1f)
                ship.mutableStats.dynamic.getStat("tahlan_KRAfterimageTracker").modifyFlat(
                    "tahlan_KRAfterimageTrackerID",
                    ship.mutableStats.dynamic.getStat("tahlan_KRAfterimageTracker").modifiedValue + amount
                )
                if (ship.mutableStats.dynamic.getStat("tahlan_KRAfterimageTracker").modifiedValue > AFTERIMAGE_THRESHOLD) {
                    CustomRender.addAfterimage(
                        ship = ship,
                        colorIn = AFTERIMAGE_COLOR,
                        duration = fadeOut
                    )
                    ship.mutableStats.dynamic.getStat("tahlan_KRAfterimageTracker").modifyFlat(
                        "tahlan_KRAfterimageTrackerID",
                        ship.mutableStats.dynamic.getStat("tahlan_KRAfterimageTracker").modifiedValue - AFTERIMAGE_THRESHOLD
                    )
                }
            } else {
                ship.mutableStats.timeMult.unmodify(ke_id)
                Global.getCombatEngine().timeMult.unmodify(ke_id)
            }
        }
    }

    //Built-in only
    override fun isApplicableToShip(ship: ShipAPI): Boolean {
        return false
    }

    override fun getDescriptionParam(index: Int, hullSize: HullSize): String? {
        if (index == 0) return Utils.txt("hmd_KassEng1")
        if (index == 1) return "" + TIME_MULT.toInt() + Utils.txt("%")
        if (index == 2) return "" + ARMOR_MALUS_FRIGATE.toInt() + Utils.txt("/") + ARMOR_MALUS_DESTROYER.toInt() + Utils.txt("/") + ARMOR_MALUS_CRUISER.toInt() + Utils.txt("/") + ARMOR_MALUS_CAPITAL.toInt()
        if (index == 3) return "" + SUPPLIES_MULT.toInt() + Utils.txt("%")
        if (index == 4) return "" + (OVERDRIVE_TRIGGER_PERCENTAGE * 100f).roundToInt() + Utils.txt("%")
        if (index == 5) return "" + OVERDRIVE_TIME_MULT.toInt() + Utils.txt("%")
        if (index == 6) return Utils.txt("hmd_KassEng2")
        return if (index == 7) Utils.txt("hmd_KassEng3") else null
    }

    companion object {
        const val ARMOR_MALUS_FRIGATE = 50f
        const val ARMOR_MALUS_DESTROYER = 100f
        const val ARMOR_MALUS_CRUISER = 200f
        const val ARMOR_MALUS_CAPITAL = 300f
        const val SUPPLIES_MULT = 50f
        const val OVERDRIVE_TRIGGER_PERCENTAGE = 0.3f
        const val OVERDRIVE_TIME_MULT = 30f
        const val TIME_MULT = 10f
        private val AFTERIMAGE_COLOR = Color(133, 126, 116, 90)
        private const val AFTERIMAGE_THRESHOLD = 0.4f
        private val OVERDRIVE_ENGINE_COLOR = Color(255, 44, 0)
        private val OVERDRIVE_GLOW_COLOR = Color(255, 120, 16)
        private val OVERDRIVE_JITTER_COLOR = Color(255, 63, 0, 20)
        private val OVERDRIVE_JITTER_UNDER_COLOR = Color(255, 63, 0, 80)
        private const val SO_MALFUNCTION_PROB = 0.03f
        private const val ke_id = "tahlan_KnightRefitID"
        val GH_HULLS = listOf(
            "tahlan_dominator_gh",
            "tahlan_enforcer_gh",
            "tahlan_onslaught_gh",
            "tahlan_monitor_gh",
            "tahlan_vendetta_gh",
            "tahlan_legion_gh",
            "tahlan_Ristreza_b",
            "tahlan_Ristreza",
            "tahlan_Exa_Pico",
            "tahlan_Vale",
            "tahlan_Izanami",
            "tahlan_Metafalica",
            "tahlan_Ristreza_g"
        )
    }
}